# FINAL ASSIGNMENT DA BBDD

El famoso portal inmobiliario idealista, con actividad en España, Andorra, Italia y Portugal, reveló hace poco un dataset con información de 2018. 

Este dataset cuenta con información relativa a las coordenadas aproximadas de los inmuebles (latitud y longitud), precios de venta de cada vivienda listada y diversas variables de características interiores. Los listados se enriquecieron con información oficial del catastro español (por ejemplo, calidad de los materiales de construcción) y otras características geográficas relevantes, como la distancia a puntos de interés urbano. 

En este caso nos vamos a centrar en el dataset con información sobre la ciudad de Madrid. 

# Primera Parte

In [3]:
# Importo las librerías
from sqlalchemy import create_engine
import pandas as pd

## Creamos la conexión

In [ ]:
# Parámetros de conexión
user     = 'root'
password = ''
#db = 'Idealista'
sqlServer = 'localhost' # debería ser IP o DNS (Domain Name System) del servidor de la base de datos

# Creamos string de conexión
connStr = "mysql+mysqlconnector://" + user + ":" + password + "@" + sqlServer  #+  "/"  + db

# print de string de conexión
print (connStr)

# Conectamos con el servidor
conn = create_engine(connStr).connect() 

## 0. Creación nueva base de datos

In [ ]:
# Creo una BD con nombre idealista
sSQL = "CREATE DATABASE Idealista"
conn.execute(sSQL)

In [6]:
# Uso la BD que acabo de crear
conn.execute("USE Idealista")

### dataset_idealista_madrid.csv

In [5]:
# Leo el archivo
inmuebles = pd.read_csv("dataset_idealista_madrid.csv", sep=',', index_col= 0)
inmuebles 

ASSETID  PERIOD     PRICE    UNITPRICE  CONSTRUCTEDAREA  \
0      A15019136831406238029  201803  126000.0  2680.851064               47   
1       A6677225905472065344  201803  235000.0  4351.851852               54   
2      A13341979748618524775  201803  373000.0  4973.333333               75   
3       A4775182175615276542  201803  284000.0  5916.666667               48   
4       A2492087730711701973  201803  228000.0  4560.000000               50   
...                      ...     ...       ...          ...              ...   
94810   A3962186799478940177  201812  347000.0  3017.391304              115   
94811   A3876164367053781729  201812  311000.0  3344.086022               93   
94812  A17768295786772492698  201803  342000.0  2826.446281              121   
94813  A15733144649359668635  201806  146000.0  2354.838710               62   
94814   A9716488392839016117  201812  334000.0  3591.397849               93   

       ROOMNUMBER  BATHNUMBER  HASTERRACE  HASLIFT  HASAIRCONDITIONING  ...  \
0               1           1           0        1                   1  ...   
1               1           1           0        0                   0  ...   
2               2           1           0        0                   1  ...   
3               1           1           0        1                   1  ...   
4               0           1           0        0                   0  ...   
...           ...         ...         ...      ...                 ...  ...   
94810           3           2           1        1                   1  ...   
94811           2           2           0        1                   1  ...   
94812           2           2           1        1                   1  ...   
94813           3           1           0        0                   0  ...   
94814           3           1           1        1                   1  ...   

       CADASTRALQUALITYID  BUILTTYPEID_1  BUILTTYPEID_2  BUILTTYPEID_3  \
0                     3.0              0              1              0   
1                     3.0              0              0              1   
2                     3.0              0              0              1   
3                     5.0              0              0              1   
4                     7.0              0              0              1   
...                   ...            ...            ...            ...   
94810                 3.0              0              0              1   
94811                 3.0              0              0              1   
94812                 3.0              0              0              1   
94813                 5.0              0              0              1   
94814                 5.0              0              0              1   

       DISTANCE_TO_CITY_CENTER  DISTANCE_TO_METRO  DISTANCE_TO_CASTELLANA  \
0                     8.058429           0.872075                6.868677   
1                     0.876369           0.116382                1.544125   
2                     0.907479           0.139109                1.608444   
3                     0.845462           0.144299                1.516166   
4                     1.250231           0.337098                1.794136   
...                        ...                ...                     ...   
94810                10.003059           0.866826                8.276950   
94811                10.198147           1.019788                8.496364   
94812                11.204027           1.883650                9.573127   
94813                 8.780692           0.171736                6.941217   
94814                11.070417           0.212377                8.799697   

       LONGITUDE   LATITUDE  AGENCY_ID  
0      -3.766933  40.362485          3  
1      -3.710725  40.422430          5  
2      -3.711571  40.422190          2  
3      -3.710440  40.422251          4  
4      -3.714340  40.408741          1  
...          ...        ...        ...  
94810  -3.

In [6]:
# Creo la tabla
inmuebles.to_sql("inmuebles", conn, index = False)

94815

In [9]:
# Compruebo que se ha creado la tabla
conn.execute("SHOW TABLES").fetchall()

[('inmuebles',)]

In [10]:
pd.read_sql("DESC inmuebles", conn)

Field       Type Null Key Default Extra
0                         ASSETID    b'text'  YES        None      
1                          PERIOD  b'bigint'  YES        None      
2                           PRICE  b'double'  YES        None      
3                       UNITPRICE  b'double'  YES        None      
4                 CONSTRUCTEDAREA  b'bigint'  YES        None      
5                      ROOMNUMBER  b'bigint'  YES        None      
6                      BATHNUMBER  b'bigint'  YES        None      
7                      HASTERRACE  b'bigint'  YES        None      
8                         HASLIFT  b'bigint'  YES        None      
9              HASAIRCONDITIONING  b'bigint'  YES        None      
10                HASPARKINGSPACE  b'bigint'  YES        None      
11  ISPARKINGSPACEINCLUDEDINPRICE  b'bigint'  YES        None      
12            HASNORTHORIENTATION  b'bigint'  YES        None      
13            HASSOUTHORIENTATION  b'bigint'  YES        None      
14             HASEASTORIENTATION  b'bigint'  YES        None      
15             HASWESTORIENTATION  b'bigint'  YES        None      
16                     HASBOXROOM  b'bigint'  YES        None      
17                    HASWARDROBE  b'bigint'  YES        None      
18                HASSWIMMINGPOOL  b'bigint'  YES        None      
19                     HASDOORMAN  b'bigint'  YES        None      
20                      HASGARDEN  b'bigint'  YES        None      
21                       ISDUPLEX  b'bigint'  YES        None      
22                       ISSTUDIO  b'bigint'  YES        None      
23                   ISINTOPFLOOR  b'bigint'  YES        None      
24               CONSTRUCTIONYEAR  b'double'  YES        None      
25            CADCONSTRUCTIONYEAR  b'bigint'  YES        None      
26            CADMAXBUILDINGFLOOR  b'bigint'  YES        None      
27               CADDWELLINGCOUNT  b'bigint'  YES        None      
28             CADASTRALQUALITYID  b'double'  YES        None      
29                  BUILTTYPEID_1  b'bigint'  YES        None      
30                  BUILTTYPEID_2  b'bigint'  YES        None      
31                  BUILTTYPEID_3  b'bigint'  YES        None      
32        DISTANCE_TO_CITY_CENTER  b'double'  YES        None      
33              DISTANCE_TO_METRO  b'double'  YES        None      
34         DISTANCE_TO_CASTELLANA  b'double'  YES        None      
35                      LONGITUDE  b'double'  YES        None      
36                       LATITUDE  b'double'  YES        None      
37                      AGENCY_ID  b'bigint'  YES        None

### agencias.csv

In [11]:
# Leo el archivo
agencias = pd.read_csv("agencias.csv", sep=',')
agencias 

AGENCY_ID          AGENCY_NAME  RATING
0          1  Agencia Madrid Plus     4.5
1          2     Madrid Viviendas     4.2
2          3   Madrid Propiedades     4.8
3          4  Madrid Casas Deluxe     4.7
4          5     Inmo Casa Madrid     4.1

In [12]:
# Creo la tabla
agencias.to_sql("agencias", conn, index = False)

5

In [13]:
# Compruebo que ambas tablas estan creadas
conn.execute("SHOW TABLES").fetchall()

[('agencias',), ('inmuebles',)]

In [14]:
pd.read_sql("DESC agencias", conn)

Field       Type Null Key Default Extra
0    AGENCY_ID  b'bigint'  YES        None      
1  AGENCY_NAME    b'text'  YES        None      
2       RATING  b'double'  YES        None

##  1. ¿Cuáles son las propiedades que tienen al menos 2 baños, un ascensor y una plaza de aparcamiento? Selecciona el ASSETID, BATHNUMBER, HASLIFT, HASPARKINGSPACE y PRICE 

In [15]:
sSQL =  """ 

SELECT assetid, bathnumber, haslift, hasparkingspace, price
FROM inmuebles
WHERE bathnumber >= 2 
    AND HASLIFT = 1 
    AND HASPARKINGSPACE = 1

"""
pd.read_sql(sSQL,conn)

assetid  bathnumber  haslift  hasparkingspace      price
0       A3694300518337702967           3        1                1  3187000.0
1      A17297754155641241826           3        1                1  1131000.0
2       A4444264986115631121           2        1                1   943000.0
3       A3581405731630543117           6        1                1  1939000.0
4       A6349776121498293972           2        1                1   369000.0
...                      ...         ...      ...              ...        ...
15994   A7034851435465452892           2        1                1   229000.0
15995  A10324853219120899544           2        1                1   293000.0
15996   A3962186799478940177           2        1                1   347000.0
15997   A3876164367053781729           2        1                1   311000.0
15998  A17768295786772492698           2        1                1   342000.0

[15999 rows x 5 columns]

## 2. ¿Cuál es el precio promedio de las propiedades con terraza? 

In [16]:
sSQL = """ 

SELECT ROUND(AVG(price),2) AS avg_price_inmuebles_con_terraza
FROM inmuebles
WHERE hasterrace = 1

"""

pd.read_sql(sSQL,conn)

avg_price_inmuebles_con_terraza
0                        432226.67

`El precio promedio de las propiedades con terraza es de 432,226.67€`


## 3. ¿Cuántas propiedades hay según el número de habitaciones? ¿Cuál es el precio medio de los apartamentos según el número de habitaciones? Ordénalo por número de habitaciones de menor a mayor

In [17]:
sSQL = """  

SELECT roomnumber, COUNT(*) AS numero_propiedades, ROUND(AVG(price),2) AS precio_promedio
FROM inmuebles
GROUP BY roomnumber
ORDER BY roomnumber

"""

pd.read_sql(sSQL,conn)

roomnumber  numero_propiedades  precio_promedio
0            0                2745        221108.93
1            1               13338        240288.12
2            2               28422        284740.52
3            3               33961        349289.18
4            4               11674        693296.90
5            5                3350       1114953.43
6            6                 789       1431944.23
7            7                 279       1442634.41
8            8                 142       1591267.61
9            9                  35       1212971.43
10          10                  24       2079083.33
11          11                  16       1799937.50
12          12                  18       1916666.67
13          13                   6       2337333.33
14          14                   5       1980000.00
15          15                   3       2164333.33
16          16                   2       2665500.00
17          18                   2       4219500.00
18          20                   2       1435000.00
19          33                   1        109000.00
20          93                   1        450000.00

## 4. ¿Cuáles son las propiedades cuyo precio es mayor que el precio promedio de todas las propiedades en el dataset? Muestra el ASSETID, PRICE y CONSTRUCTEDAREA

In [18]:
sSQL = """  

SELECT assetid, price, constructedarea
FROM inmuebles 
WHERE price > (SELECT AVG(price) from inmuebles)

"""

pd.read_sql(sSQL,conn)

assetid      price  constructedarea
0      A18372428154681111419   498000.0              127
1       A9587449507628658013   425000.0               70
2       A3694300518337702967  3187000.0              360
3      A11532129615377606790   422000.0              114
4       A2893199643597593313  1013000.0              155
...                      ...        ...              ...
29694   A3107360537414270361   936000.0              205
29695  A10395827789147723674   460000.0              115
29696  A12819204236902632452   460000.0               93
29697  A13569164165299053978   457000.0               93
29698   A7764751834904574987   426000.0               94

[29699 rows x 3 columns]

##  5. ¿Cuántas propiedades tienen más de 3 habitaciones y están situadas en la última planta del edificio? 

In [19]:
sSQL = """ 

SELECT COUNT(*) AS conteo
FROM inmuebles
WHERE roomnumber > 3
     AND isintopfloor = 1

"""

pd.read_sql(sSQL,conn)

conteo
0     356

`Hay 356 propiedades situadas en la última planta con más de 3 habitaciones`

## 6. ¿Cuál es la propiedad más cara registrada en el dataset, es decir, su ASSETID?

In [20]:
sSQL = """  

SELECT assetid AS propiedad_mas_cara
FROM inmuebles
WHERE price = (SELECT MAX(price) from inmuebles)

"""

pd.read_sql(sSQL,conn)

propiedad_mas_cara
0  A17400384699956653141

`La propiedad más cara registrada es la propiedad con id: A17400384699956653141`

## 7. ¿Cuál es la distancia promedio al centro de la ciudad para las propiedades construidas después del año 2000?

Uso el año de construcción según los registros catastrales porque el año de construcción del inmueble contiene valores nulos

In [22]:
sSQL = """ 

SELECT ROUND(AVG(distance_to_city_center),2) AS avg_distancia_centro
FROM inmuebles
WHERE cadconstructionyear > 2000

"""

pd.read_sql(sSQL,conn)

avg_distancia_centro
0                   6.9

`La distancia promedio al centro de la ciudad para las propiedades construidas después del año 2000 es de 6.90 km`

## 8. ¿Cuántas propiedades tienen piscina, jardín y plaza de aparcamiento? 

In [23]:
sSQL = """  

SELECT COUNT(assetid) AS conteo_inmuebles
FROM inmuebles
WHERE hasswimmingpool = 1
    AND hasgarden = 1
    AND hasparkingspace = 1

"""
pd.read_sql(sSQL,conn)

conteo_inmuebles
0              9034

`Hay 9034 inmuebles que tienen piscina, jardín y aparcamiento`

## 9. ¿Qué porcentaje de propiedades de segunda mano están en buen estado y tienen más de 2 baños? 

Divido las propiedades de segunda mano en buen estado con más de 2 baños, entre las propiedades de segunda mano totales(propiedades para restaurar y en buen estado).

In [24]:
sSQL = """


SELECT
    ROUND(SUM(CASE WHEN builttypeid_3 = 1 AND bathnumber > 2 THEN 1 ELSE 0 END) * 100 / 
    COUNT(CASE WHEN builttypeid_2 = 1 OR builttypeid_3 = 1 THEN 1 END),2) AS porcentaje_inmuebles_buen_estado_con_2_baños
FROM inmuebles

"""

pd.read_sql(sSQL, conn)

porcentaje_inmuebles_buen_estado_con_2_baños
0                                          7.61

`Un 7.61% de propiedades de segunda mano están en buen estado y tienen más de 2 baños`

## 10. ¿Hay alguna tendencia en el valor medio de las propiedades en los distintos meses del año 2018? ¿Y en el valor máximo o mínimo? 

In [25]:
sSQL = """

SELECT period, AVG(price) AS valor_medio, MAX(price) AS valor_max, MIN(price) AS valor_min
FROM inmuebles
GROUP BY period
ORDER BY period

"""

pd.read_sql(sSQL,conn)

period    valor_medio  valor_max  valor_min
0  201803  404960.766423  6729000.0    32000.0
1  201806  367527.505533  7138000.0    21000.0
2  201809  365467.225944  6829000.0    24000.0
3  201812  410952.654168  8133000.0    24000.0

- `En cuanto al valor medio, al principio y final de año es un valor superior, a medida que se acerca el verano baja`
- `El valor máximo también tiene su máximo a final de año`
- `El valor mínimo tiene su máximo a principio de año, se reduce al entrar junio, y finalmente se recupera un poco`

## 11. Supón que quieres aumentar en un 5% el precio de todas las propiedades que tienen aire acondicionado. Haz una actualización del campo PRICE para reflejar este cambio. 

In [26]:
sSQL = """

UPDATE inmuebles
SET price = price * 1.05
WHERE hasairconditioning = 1

"""

conn.execute(sSQL)

## 12. Elimina todas las propiedades que tengan menos de 30 metros cuadrados de superficie construida. 

In [27]:
sSQL = """

DELETE FROM inmuebles
WHERE constructedarea < 30

"""

conn.execute(sSQL)

## 13. Crea una vista que muestre solo las propiedades con terraza y piscina, incluyendo los campos ASSETID, PRICE, ROOMNUMBER y BATHNUMBER. 

In [28]:
sSQL = """
CREATE VIEW propiedadesTerrazaPiscina AS 
SELECT assetid, price, roomnumber, bathnumber
FROM inmuebles
WHERE hasterrace = 1
    AND hasswimmingpool = 1

"""

conn.execute(sSQL)

## 14. ¿Cuál es el precio promedio por metro cuadrado para propiedades con orientación sur, este y oeste?

In [29]:
sSQL = """

SELECT ROUND(AVG(unitprice),2) as avg_precio_m2
FROM inmuebles
WHERE hassouthorientation = 1 
    AND haswestorientation = 1
    AND haseastorientation = 1

"""
pd.read_sql(sSQL,conn)

avg_precio_m2
0        4322.46

`El precio promedio por metro cuadrado de las propiedades con orientación al sur, este y oeste es 4322,46€/m2`

## 15. ¿Cuál es la diferencia de los precios promedios entre propiedades que incluyen la plaza de aparcamiento y las que no? 

In [12]:
sSQL = """
SELECT 
    AVG(CASE WHEN HASPARKINGSPACE = 1 THEN PRICE END) as parking,
    AVG(CASE WHEN HASPARKINGSPACE = 0 THEN PRICE END) as sin_parking,
    ABS(
        AVG(CASE WHEN HASPARKINGSPACE = 1 THEN PRICE END) - 
        AVG(CASE WHEN HASPARKINGSPACE = 0 THEN PRICE END)
    ) as price_difference
FROM inmuebles
"""
pd.read_sql(sSQL, conn)


parking    sin_parking  price_difference
0  584696.555118  357555.371269     227141.183849

`La diferencia en el precio de los inmuebles con parking es de 227.141,18 €`


# Segunda Parte

In [31]:
# Veo la tabla agencias
pd.read_sql("SELECT * FROM agencias", conn)

AGENCY_ID          AGENCY_NAME  RATING
0          1  Agencia Madrid Plus     4.5
1          2     Madrid Viviendas     4.2
2          3   Madrid Propiedades     4.8
3          4  Madrid Casas Deluxe     4.7
4          5     Inmo Casa Madrid     4.1

## 1. ¿Qué propiedades gestionadas por agencias tienen piscina y plaza de aparcamiento? 

In [32]:
sSQL = """  

SELECT assetid
FROM inmuebles
INNER JOIN agencias
ON inmuebles.agency_id = agencias.agency_id
WHERE  hasswimmingpool = 1
    AND hasparkingspace = 1
"""

pd.read_sql(sSQL, conn)

assetid
0       A4671190715338907020
1      A13321527061841065076
2       A7976691527475320095
3      A14703419175734607889
4       A3347762916822626363
...                      ...
10827  A10324853219120899544
10828   A8974485819301533463
10829   A3962186799478940177
10830   A3876164367053781729
10831  A17768295786772492698

[10832 rows x 1 columns]

## 2. ¿Cuáles son todas las propiedades gestionadas por una agencia con puntuación mayor a 4.5? Muestra el ASSETID, AGENCY_NAME, PRICE y RATING. 

In [33]:
sSQL = """  

SELECT assetid, agency_name, price, rating
FROM inmuebles
INNER JOIN agencias
ON inmuebles.agency_id = agencias.agency_id
WHERE agencias.rating > 4.5

"""

pd.read_sql(sSQL, conn)

assetid          agency_name     price  rating
0      A15019136831406238029   Madrid Propiedades  132300.0     4.8
1       A4775182175615276542  Madrid Casas Deluxe  298200.0     4.7
2      A18372428154681111419   Madrid Propiedades  498000.0     4.8
3       A9587449507628658013  Madrid Casas Deluxe  425000.0     4.7
4      A15416265623565967313   Madrid Propiedades  211000.0     4.8
...                      ...                  ...       ...     ...
37793   A8974485819301533463  Madrid Casas Deluxe  205800.0     4.7
37794   A8870664464301737677   Madrid Propiedades  354900.0     4.8
37795   A3876164367053781729  Madrid Casas Deluxe  326550.0     4.7
37796  A17768295786772492698  Madrid Casas Deluxe  359100.0     4.7
37797   A9716488392839016117  Madrid Casas Deluxe  350700.0     4.7

[37798 rows x 4 columns]

## 3. ¿Qué inmobiliarias no tienen propiedades listadas en el dataset? 

In [34]:
sSQL = """ 

SELECT agency_name
FROM inmuebles
RIGHT JOIN agencias
ON inmuebles.agency_id = agencias.agency_id
WHERE inmuebles.agency_id IS NULL
"""

pd.read_sql(sSQL, conn)

Empty DataFrame
Columns: [agency_name]
Index: []

`No hay ninguna inmobiliaria sin propiedades listadas`

## 4. ¿Cuál es la inmobiliaria con el precio promedio más alto en propiedades listadas? 

In [35]:
sSQL = """ 

WITH listed_properties AS (
SELECT agency_name, AVG(price) AS avg_price
FROM inmuebles
INNER JOIN agencias
ON inmuebles.agency_id = agencias.agency_id
GROUP BY agency_name
)

SELECT agency_name
FROM listed_properties
WHERE avg_price = (SELECT MAX(avg_price) FROM listed_properties)
"""

pd.read_sql(sSQL, conn)

agency_name
0  Inmo Casa Madrid

`La inmobiliaria con el precio promedio más alto es Inmo Casa Madrid`

## 5. Muestra todos los inmuebles y, si existe, el nombre de su agencia. 

Uso el LEFT JOIN para que seleccione los assetid aunque no tengan ninguna agencia

In [36]:
sSQL = """  

SELECT assetid, agency_name, price, rating
FROM inmuebles
LEFT JOIN agencias
ON inmuebles.agency_id = agencias.agency_id

"""

pd.read_sql(sSQL, conn)

assetid          agency_name     price  rating
0      A15019136831406238029   Madrid Propiedades  132300.0     4.8
1       A6677225905472065344     Inmo Casa Madrid  235000.0     4.1
2      A13341979748618524775     Madrid Viviendas  391650.0     4.2
3       A4775182175615276542  Madrid Casas Deluxe  298200.0     4.7
4       A2492087730711701973  Agencia Madrid Plus  228000.0     4.5
...                      ...                  ...       ...     ...
93893   A3962186799478940177     Madrid Viviendas  364350.0     4.2
93894   A3876164367053781729  Madrid Casas Deluxe  326550.0     4.7
93895  A17768295786772492698  Madrid Casas Deluxe  359100.0     4.7
93896  A15733144649359668635  Agencia Madrid Plus  146000.0     4.5
93897   A9716488392839016117  Madrid Casas Deluxe  350700.0     4.7

[93898 rows x 4 columns]

## 6. ¿Cuántos inmuebles gestiona cada agencia? Ordénalo de mayor a menor. 

In [37]:
sSQL = """  

SELECT agency_name, COUNT(assetid) as viviendas_gestionadas
FROM inmuebles
INNER JOIN agencias
ON inmuebles.agency_id = agencias.agency_id
GROUP BY agency_name
ORDER BY COUNT(assetid) DESC

"""

pd.read_sql(sSQL, conn)

agency_name  viviendas_gestionadas
0   Madrid Propiedades                  19042
1  Agencia Madrid Plus                  18955
2  Madrid Casas Deluxe                  18756
3     Inmo Casa Madrid                  18584
4     Madrid Viviendas                  18561

In [38]:
conn.close()